dashboard/lib/sample_data.rb

```diff
@@ -0,0 +1,199 @@
+include FactoryGirl::Syntax::Methods
+
+class SampleData
+  SAMPLE_TEACHER_EMAIL = 'testteacher@code.org'.freeze
+  SAMPLE_TEACHER_PASSWORD = '00secret'.freeze
+  SAMPLE_TEACHER_NAME = 'TestTeacher Codeberg'.freeze
+  SAMPLE_STUDENT_NAME_FORMAT = 'TestStudent%s Codeberg'.freeze
+  SAMPLE_STUDENT_NAME_REGEX = /TestStudent\d* Codeberg/
+
+  @@rng = nil
+
+  # Returns a seeded random number generator for consistent test data
+  def self.rng
+    @@rng ||= Random.new(0)
+  end
+
+  # Creates sample data
+  def self.seed
+    raise "Should not be run outside of development" unless CDO.rack_env?(:development)
+
+    # Create a test teacher
+    teacher = create_teacher SAMPLE_TEACHER_EMAIL, SAMPLE_TEACHER_PASSWORD,
+      SAMPLE_TEACHER_NAME
+
+    # Create normal-sized and large sections for each of CSF, CSD and CSP
+
+    create_section(
+      teacher: teacher, name: 'CSF 1',
+      login_type: Section::LOGIN_TYPE_PICTURE, grade: 2, age_min: 7,
+      age_max_inclusive: 9, script_name: 'coursea-2018', num_students: 30,
+      use_imperfect_results: true
+    )
+
+    create_section(
+      teacher: teacher, name: 'CSF 2 (Large)',
+      login_type: Section::LOGIN_TYPE_PICTURE, grade: 2, age_min: 7,
+      age_max_inclusive: 9, script_name: 'coursea-2018', num_students: 150,
+      use_imperfect_results: true
+    )
+
+    create_section(
+      teacher: teacher, name: 'CSD 1',
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 7, age_min: 13,
+      age_max_inclusive: 14, script_name: 'csd1-2018', num_students: 30,
+      use_imperfect_results: false
+    )
+
+    create_section(
+      teacher: teacher, name: 'CSD 2 (Large)',
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 7, age_min: 13,
+      age_max_inclusive: 14, script_name: 'csd1-2018', num_students: 150,
+      use_imperfect_results: false
+    )
+
+    create_section(
+      teacher: teacher, name: 'CSP 1',
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 10, age_min: 14,
+      age_max_inclusive: 16, script_name: 'csp1-2018', num_students: 30,
+      use_imperfect_results: false
+    )
+
+    create_section(
+      teacher: teacher, name: 'CSP 2 (Large)',
+      login_type: Section::LOGIN_TYPE_EMAIL, grade: 10, age_min: 14,
+      age_max_inclusive: 16, script_name: 'csp1-2018', num_students: 150,
+      use_imperfect_results: false
+    )
+  end
+
+  # Hard-delete the teacher and all of the teacher's sections and students
+  # and recreate. Sections and followers would be soft-deleted by
+  # dependency when we delete the teacher; but to not leave a trail of
+  # old test data behind, we explictly hard-delete.
+  def self.create_teacher(email, password, name)
```

Comment:
Since deleting happens in this method that _could_ conceivably in the future be called from anywhere, how about duplicating the
```rb
raise "Should not be run outside of development" unless CDO.rack_env?(:development)
```
check at the top of this method too, just to be paranoid?

Reply:
Oh, I see you've done this down on line 82.  I'd pull it up here - as currently written, you could hit the hard-delete on a section without raising if that section had no students.

Reply:
Thanks. Added another one at top of method. I'd like to have defense in depth as the code changes in the future, perhaps inadvertently gets copied/moved into other contexts, etc.